File: frontend/src/app/services/competition.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { CompetitionResponse } from '../model/dto/response/CompetitionResponse';
import { UpdateCompetitionRequest } from '../model/dto/request/UpdateCompetitionRequest';
import { CompetitorResponse } from '../model/dto/response/CompetitorResponse';
import * as myGlobals from 'src/globals';
import { CompetitorRequest } from '../model/dto/request/CompetitorRequest';

@Injectable({
  providedIn: 'root',
})
export class CompetitionService {
  private path = myGlobals.path + '/competition';

  constructor(private http: HttpClient, private router: Router) {
    this.router.routeReuseStrategy.shouldReuseRoute = () => false;
  }

  getCompetitionNames(competitionIds: number[]) {
    var stringIds: string = '';
    for (let i = 0; i < competitionIds.length; i++) {
      stringIds += competitionIds[i] + '&';
    }
    return this.http.get<CompetitionResponse[]>(
      this.path + '/names/' + stringIds
    );
  }

  getSearchResults(query: string) {
    return this.http.get<CompetitionResponse[]>(this.path + '/search/' + query);
  }

  getCompetition(id: number) {
    return this.http.get<CompetitionResponse>(this.path + '/' + id);
  }

  updateCompetition(request: UpdateCompetitionRequest) {
    return this.http.put<CompetitionResponse>(this.path, request);
  }

  updateNames(competitorDto: CompetitorRequest[], competitionId: number) {
    return this.http.put(
      this.path + '/competitor/names/' + competitionId,
      competitorDto
    );
  }

  getCompetitors(competitionId: number) {
    return this.http.get<CompetitorResponse[]>(
      this.path + '/competitor/' + competitionId
    );
  }
}

File: backend/src/main/java/com/lgm/demo/service/CompetitionService.java
package com.lgm.demo.service;

import com.lgm.demo.model.Competition;
import com.lgm.demo.model.dto.request.CompetitorRequest;
import com.lgm.demo.model.dto.request.UpdateCompetitionRequest;
import com.lgm.demo.model.dto.response.CompetitionResponse;
import com.lgm.demo.model.dto.response.CompetitorResponse;

import java.util.List;

public interface CompetitionService{
    Competition getCompetition(Long competitionId);
    List<CompetitionResponse> getCompetitionNames(String competitionIdsString);
    List<CompetitionResponse> getSearchResults(String query);
    CompetitionResponse updateCompetition(UpdateCompetitionRequest request);
    void updateCompetitorNames(List<CompetitorRequest> request, Long competitionId);
    List<CompetitorResponse> getCompetitors(Long competitionId);
}
